### PR TITLE
Restructured test suites

### LINKED ISSUE #4: Move tests to mongo-knex/ngl-lang accordingly
This package is just an interface that connects ngl-lang and mongo-knex together. Most of the test in the test suite should be moved closer to the library that is responsible for a particular part of logic instead of testing it here. 
 
#refs https://github.com/NexesJS/mongo-knex/issues/3

### PR DESCRIPTION
**Pull request by gargol**
closes #4 

- Removed much of redundant test suites
- Cleaned up main module api suite 

Test suites for mongo-knex and mingo were removed as packages by themselves prove the validity of their code. We just need to know if NQL glues them together properly, which is proved by remaining suites

### CODE REVIEW

**Comment by gargol** [comment]
Left the suite only with the basic module api test. The rest is just too specific to what mongo-knex or mingo are supposed to do by themselves.

**Comment by kirrg001** [comment]
Are the removed tests covered in nql-lang?

**Comment by kirrg001** [comment]
I think this needs to stay

> https://github.com/NexesJS/NQL/pull/5/commits/0ce5ffc7f61991c9fb62025248ce847a48f58de9

Because mingo is used in NQL. There is no repository.

**Comment by gargol** [comment]
> Are the removed tests covered in nql-lang?

yes, the suite is a lot more extensive there - https://github.com/NexesJS/NQL-Lang/blob/master/test/parser.test.js#L16

> I think this needs to stay 0ce5ffc

I was thinking if it should be removed completely, partially or stay as is. But decided to go with full removal because to me it seems like it's just testing if the result produced by nql-lang parser is applicable to json objects. The correctness of ngl-lang output is covered by the modules unit test, the correctness of mingo using query operators is covered by the module itself [quite well](https://github.com/kofrasa/mingo/blob/master/test/query_operators.js). In my opinion,  the only thing left to cover if ngl 'glues' nql-lang and mingo correctly which is done by [api.test.js](https://github.com/NexesJS/NQL/blob/master/test/api.test.js). @kirrg001 what's your perspective on this?

**Comment by kirrg001** [comment]
> yes, the suite is a lot more extensive there - https://github.com/NexesJS/NQL-Lang/blob/master/test/parser.test.js#L16

Cool. Did you compare both and check for missing cases which were covered in this repo, but not covered in nql-lang?

**Comment by kirrg001** [comment]
> I was thinking if it should be removed completely, partially or stay as is. But decided to go with full removal because to me it seems like it's just testing if the result produced by nql-lang parser is applicable to json objects. The correctness of ngl-lang output is covered by the modules unit test, the correctness of mingo using query operators is covered by the module itself quite well. In my opinion, the only thing left to cover if ngl 'glues' nql-lang and mingo correctly which is done by api.test.js. @kirrg001 what's your perspective on this?

Yeah makes sense. 
But we said we want to keep one test per use case or?

filter -> nql -> nql-lang -> mingo
filter -> nql -> nql-lang -> mongo-knex

**Comment by gargol** [comment]
> But we said we want to keep one test per use case or?

We didn't come to final decision, but maybe having just one simple case  for mingo and mongo-knex would be good. With a comment that it's just a sanity check if things are guled together correctly and more extended logic is meant to be tested in those modules internally.

> filter -> nql -> nql-lang -> mingo
filter -> nql -> nql-lang -> mongo-knex

yes :100: the path which takes for filter to be transformed into JSON or SQL statement.

**Comment by kirrg001** [comment]
> We didn't come to final decision, but maybe having just one simple case for mingo and mongo-knex would be good. With a comment that it's just a sanity check if things are guled together correctly and more extended logic is meant to be tested in those modules internally.

Yeah. I think we should not remove everything and leave one test per use case in this PR?

**Comment by gargol** [comment]
> Cool. Did you compare both and check for missing cases which were covered in this repo, but not covered in nql-lang?

I didn't check assertion by assertion, but each `describe` group was covered 100% by nql-lang and looked even more exntensive than what was availalbe in here. Same for mongo-knex, hence
the link in first commit message - https://github.com/NexesJS/mongo-knex/blob/master/test/unit/convertor.test.js :smiley:

**Comment by gargol** [comment]
@kirrg001 done the changes discussed above ;)

File: test/nql_knex.test.js [hunk 1]
```diff
@@ -12,254 +12,13 @@ const makeQuery = (nqlString, options) => {
     const postKnex = knex('posts');
     const filter = nql.parse(nqlString, options);
 
-    // console.log('Filter', require('util').inspect(filter, false, null)); // eslint-disable-line no-console
-
     return mongoToKnex(postKnex, filter, {});
```
Comment (kirrg001) [inline test/nql_knex.test.js]: wondering why the integration tests call the sub dependency manually?
instead of using the API of NQL?

Comment (gargol) [reply test/nql_knex.test.js]: Looked into it a bit more and the main api suite covers more than this test at the moment - https://github.com/NexesJS/NQL/blob/master/test/api.test.js#L35. Think we'd need something similar integration tests in mongo-knex here to check the base case, and the test would then look more like:
```
    it('should match based on simple id', () => {
        // set up the db first

        const query = nql('id:3');

        return query
            .querySQL(knex('posts'))
            .select()
            .then((result) => {
                result.should.be.an.Array().with.lengthOf(3);
            });
    });
```
Should I go on and implement this ^ ? What are your thought @kirrg001 ?

Comment (kirrg001) [reply test/nql_knex.test.js]: 👍

**Comment by gargol** [comment]
@kirrg001 last commit is mostly a copy of mongo-knex setup to be able to test [this way](https://github.com/NexesJS/NQL/pull/5#discussion_r232239194)

File: test/integration/knex/schema.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+const Promise = require('bluebird');
+const TABLES = ['posts_tags', 'posts', 'tags', 'users'];
```
Comment (kirrg001) [inline test/integration/knex/schema.js]: we only use a basic table in the test suite

File: test/integration/knex/schema.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+module.exports.up = function (knex) {
+    // Before all tests, we load any base data (data that won't change)
+    return Promise.resolve()
+        .then(() => knex.schema.createTable('posts', (table) => {
+            table.increments('id').primary();
+            table.string('title', 191).defaultTo('(Untitled)');
+            table.boolean('featured').defaultsTo(false);
+            table.string('image', 191).nullable();
+            table.string('status', 191).nullable();
+            table.integer('author_id').unsigned().references('users.id');
```
Comment (kirrg001) [inline test/integration/knex/schema.js]: there is no users table. this would fail on MySQL?

File: test/integration/nql_knex.test.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+const knex = utils.db.client;
+
+/**
+ * The purpose of this file is to prove that NQL
+ * is not just transformed to mongo queries correctly
+ * but that this can be used in real world settings to query SQL databases
+ */
+
+describe.skip('Integration with Knex', () => {
```
Comment (kirrg001) [inline test/integration/nql_knex.test.js]: why is this skipped?

Comment (gargol) [reply test/integration/nql_knex.test.js]: my fault, committed this too quick :see_no_evil:

**Comment by kirrg001** [comment]
I'd suggest to ask John tomorrow to enable travis for NQL. Then we can run this PR on travis.

**Comment by gargol** [comment]
@kirrg001 linting is failing with a ton of errors, do we wanna skip it for now?

**Comment by kirrg001** [comment]
> @kirrg001 linting is failing with a ton of errors, do we wanna skip it for now?

Did you bump eslint plugin to the latest?

**Comment by gargol** [comment]
> Did you bump eslint plugin to the latest?
Nope. Will try it now

**Comment by gargol** [comment]
`yarn test` passes nicely, it's good to be integrated with Travis now. Also tested MySQL version, it works.

**Comment by gargol** [comment]
@kirrg001 rebased upon master, lmk if we can :ship:  this

**Comment by kirrg001** [comment]
@gargol up to you. either we merge it now or tomorrow when travis is enabled 🤘🏽👻

**Comment by gargol** [comment]
Would do travis as a separate PR. Don't want to deal with rebases too much :smile_cat:

**Comment by kirrg001** [comment]
Yeah 100% separate PR. Just wanted to see if the PR passes. But feel free to merge this already :) Doesn't rly matter

**Comment by ErisDS** [comment]
> We didn't come to final decision, but maybe having just one simple case for mingo and mongo-knex would be good. With a comment that it's just a sanity check if things are guled together correctly and more extended logic is meant to be tested in those modules internally.

I have only just seen this PR, and disagree strongly with the decision made here. Mingo is a 3rd party module that we depend on heavily for functionality not just here in NQL but deep within Ghost.

Now, when a new version of mingo comes along, we have no way to determine if our use cases are still handled correctly. That's the purpose of integration tests.

The tests were already written, and took 146ms to run. What was the reason for suggesting their removal in the first place?

File: config/nconf.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+const Nconf = require('nconf');
```
[the remark this answers is not in the document]
Comment (gargol) [reply config/nconf.js]: Good point :+1: Was following the structure in [mongo-knex](https://github.com/NexesJS/mongo-knex/tree/master/config). Will change it in both places

**Comment by gargol** [comment]
Can't find any specific reference point around decision making for this PR except the [issue](https://github.com/NexesJS/NQL/issues/4) it reference. The idea behind removing test suites from this repo was to leave the testing of implementation to the libraries that this package glues together to those libraries. For example, nql_mingo.test.js suite was removed as it seemed to have tested what mingo itself [tests](https://github.com/kofrasa/mingo/blob/master/test/query_operators.js) - a compliance with MongoDB query language. 
I see the point of not having visibility over whether the version bump is breaking any of our use cases. We could trust that mingo's implementation is following MongoDB query language spec 100%, but that's probably a very bad idea with external dependency we don't have any control over.

**Comment by ErisDS** [comment]
What I see is, the issue makes a proposal without reasoning. All further discussion has been around details. 

I can't see that the tests were getting in the way or causing any problems, so I'm curious about the justification for spending time on removing them?

**Comment by gargol** [comment]
I think I jumped to fast into a [conclusion](https://github.com/NexesJS/NQL/pull/5#issuecomment-437032016) of removing the test suite and missed it's of it being here in the first place. Also made a mistake of not documenting properly such a decision. Will get the integration suite for mingo back to life. Should the suite for nql_knex.test.js stay removed? As it's well covered in [here](https://github.com/NexesJS/mongo-knex/blob/master/test/unit/convertor.test.js#L20)

**Comment by ErisDS** [comment]
My point is more that I don't understand why the issue was even raised, not just the conclusion. I don't see any value to having removed any test suites unless they were explicitly causing you problems whilst doing other important work.

Right now, the focus needs to be on finishing up the tasks which contribute to the goal of swapping out GQL for NQL.

**Pull request merged**